src/java.base/windows/native/libjava/WinNTFileSystem_md.c

```diff
@@ -649,31 +649,41 @@ Java_java_io_WinNTFileSystem_createFileExclusively0(JNIEnv *env, jclass cls,
 }
 
 static int
-removeFileOrDirectory(const jchar *path)
+removeFileOrDirectory(const jchar *path, jboolean deleteReadOnly)
 {
     /* Returns 0 on success */
-    DWORD a;
-
-    SetFileAttributesW(path, FILE_ATTRIBUTE_NORMAL);
-    a = GetFileAttributesW(path);
+    DWORD a = GetFileAttributesW(path);
     if (a == INVALID_FILE_ATTRIBUTES) {
         return 1;
     } else if (a & FILE_ATTRIBUTE_DIRECTORY) {
+        // read-only attribute cannot be set on directories
         return !RemoveDirectoryW(path);
     } else {
-        return !DeleteFileW(path);
+        // unset read-only attribute if deleting read-only files is enabled
+        BOOL attrSet = FALSE;
+        if (deleteReadOnly && ((a & FILE_ATTRIBUTE_READONLY) != 0))
+            attrSet = SetFileAttributesW(path, a & (~FILE_ATTRIBUTE_READONLY));
```

Comment:
A small comment on the name "attrSet". It hints that it will be set to TRUE if the attribute is set whereas the use here is to mean that SetFileAttributesW has succeeded, a bit confusing because it's clearing the READONLY value, not setting it.  If you rename to something that has the word "modified" or "cleared" might be easy for future maintainers.

Reply:
Name updated in efca3cf.